File: client-android/serenada-core/src/main/java/app/serenada/core/call/StatsPoller.kt
package app.serenada.core.call

import android.os.Handler
import java.util.concurrent.ExecutorService

class StatsPoller(
    private val handler: Handler,
    private val clock: SessionClock,
    private val statsExecutorProvider: () -> ExecutorService?,
    private val isActivePhase: () -> Boolean,
    private val getPeerSlots: () -> List<PeerConnectionSlotProtocol>,
    private val onStatsUpdated: (RealtimeCallStats) -> Unit,
    private val onRefreshRemoteParticipants: () -> Unit,
) {
    private var remoteVideoStatePollRunnable: Runnable? = null
    private var webrtcStatsRequestInFlight = false
    private var lastWebRtcStatsPollAtMs = 0L

    fun start() {
        if (remoteVideoStatePollRunnable != null) return
        val runnable = object : Runnable {
            override fun run() {
                onRefreshRemoteParticipants()
                pollWebRtcStats()
                handler.postDelayed(this, 500)
            }
        }
        remoteVideoStatePollRunnable = runnable
        handler.post(runnable)
    }

    fun stop() {
        remoteVideoStatePollRunnable?.let { handler.removeCallbacks(it) }
        remoteVideoStatePollRunnable = null
        webrtcStatsRequestInFlight = false
        lastWebRtcStatsPollAtMs = 0L
    }

    private fun pollWebRtcStats() {
        if (!isActivePhase()) return
        val now = clock.nowMs()
        if (webrtcStatsRequestInFlight) return
        if (now - lastWebRtcStatsPollAtMs < WEBRTC_STATS_POLL_INTERVAL_MS) return
        val slots = getPeerSlots()
        if (slots.isEmpty()) return
        webrtcStatsRequestInFlight = true
        val executor = statsExecutorProvider()?.takeIf { !it.isShutdown }
        if (executor == null) { webrtcStatsRequestInFlight = false; return }
        try {
            executor.execute {
                val stats = mutableListOf<RealtimeCallStats>()
                var remaining = slots.size
                slots.forEach { slot ->
                    slot.collectWebRtcStats { _, realtimeStats ->
                        synchronized(stats) {
                            realtimeStats?.let(stats::add)
                            remaining -= 1
                            if (remaining == 0) {
                                val merged = mergeRealtimeStats(stats)
                                handler.post {
                                    webrtcStatsRequestInFlight = false
                                    lastWebRtcStatsPollAtMs = clock.nowMs()
                                    if (merged != null) {
                                        onStatsUpdated(merged)
                                    }
                                }
                            }
                        }
                    }
                }
            }
        } catch (_: java.util.concurrent.RejectedExecutionException) {
            webrtcStatsRequestInFlight = false
        }
    }

    companion object {
        private const val WEBRTC_STATS_POLL_INTERVAL_MS = 2000L

        fun mergeRealtimeStats(stats: List<RealtimeCallStats>): RealtimeCallStats? {
            if (stats.isEmpty()) return null
            fun sumN(sel: (RealtimeCallStats) -> Double?) = stats.mapNotNull(sel).sum().takeIf { stats.any { s -> sel(s) != null } }
            fun maxN(sel: (RealtimeCallStats) -> Double?) = stats.mapNotNull(sel).maxOrNull()
            fun minN(sel: (RealtimeCallStats) -> Double?) = stats.mapNotNull(sel).minOrNull()
            return RealtimeCallStats(
                transportPath = stats.mapNotNull { it.transportPath }.distinct().joinToString().ifBlank { null },
                rttMs = maxN { it.rttMs }, availableOutgoingKbps = minN { it.availableOutgoingKbps },
                audioRxPacketLossPct = maxN { it.audioRxPacketLossPct }, audioTxPacketLossPct = maxN { it.audioTxPacketLossPct },
                audioJitterMs = maxN { it.audioJitterMs }, audioPlayoutDelayMs = maxN { it.audioPlayoutDelayMs },
                audioConcealedPct = maxN { it.audioConcealedPct },
                audioRxKbps = sumN { it.audioRxKbps }, audioTxKbps = sumN { it.audioTxKbps },
                videoRxPacketLossPct = maxN { it.videoRxPacketLossPct }, videoTxPacketLossPct = maxN { it.videoTxPacketLossPct },
                videoRxKbps = sumN { it.videoRxKbps }, videoTxKbps = sumN { it.videoTxKbps },
                videoFps = maxN { it.videoFps }, videoResolution = stats.asReversed().firstNotNullOfOrNull { it.videoResolution },
                videoFreezeCount60s = stats.mapNotNull { it.videoFreezeCount60s }.sum().takeIf { it > 0 },
                videoFreezeDuration60s = sumN { it.videoFreezeDuration60s },
                videoRetransmitPct = maxN { it.videoRetransmitPct }, videoNackPerMin = sumN { it.videoNackPerMin },
                videoPliPerMin = sumN { it.videoPliPerMin }, videoFirPerMin = sumN { it.videoFirPerMin },
                updatedAtMs = stats.maxOf { it.updatedAtMs },
            )
        }
    }
}

File: client-ios/SerenadaCore/Sources/Call/StatsPoller.swift
import Combine
import Foundation

@MainActor
final class StatsPoller {
    private let clock: SessionClock
    private let isActivePhase: () -> Bool
    private let getPeerSlots: () -> [any PeerConnectionSlotProtocol]
    private let onStatsUpdated: (RealtimeCallStats) -> Void
    private let onRefreshRemoteParticipants: () -> Void

    private var pollTimerCancellable: AnyCancellable?
    private var webrtcStatsRequestInFlight = false
    private var lastWebRtcStatsPollAtMs: Int64 = 0

    init(
        clock: SessionClock,
        isActivePhase: @escaping () -> Bool,
        getPeerSlots: @escaping () -> [any PeerConnectionSlotProtocol],
        onStatsUpdated: @escaping (RealtimeCallStats) -> Void,
        onRefreshRemoteParticipants: @escaping () -> Void
    ) {
        self.clock = clock
        self.isActivePhase = isActivePhase
        self.getPeerSlots = getPeerSlots
        self.onStatsUpdated = onStatsUpdated
        self.onRefreshRemoteParticipants = onRefreshRemoteParticipants
    }

    func start() {
        stop()

        pollTimerCancellable = clock.scheduleRepeating(intervalSeconds: 0.5) { [weak self] in
            guard let self else { return }
            self.onRefreshRemoteParticipants()
            self.pollWebRtcStats()
        }
    }

    func stop() {
        pollTimerCancellable?.cancel()
        pollTimerCancellable = nil
        webrtcStatsRequestInFlight = false
        lastWebRtcStatsPollAtMs = 0
    }

    private func pollWebRtcStats() {
        guard isActivePhase() else { return }

        let now = clock.nowMs()
        if webrtcStatsRequestInFlight { return }
        if now - lastWebRtcStatsPollAtMs < 2000 { return }

        webrtcStatsRequestInFlight = true

        let slots = getPeerSlots()
        guard !slots.isEmpty else {
            webrtcStatsRequestInFlight = false
            lastWebRtcStatsPollAtMs = now
            onStatsUpdated(.empty)
            return
        }

        let group = DispatchGroup()
        var stats: [RealtimeCallStats] = []
        let lock = NSLock()

        for slot in slots {
            group.enter()
            slot.collectRealtimeCallStats { realtimeStats in
                lock.lock()
                stats.append(realtimeStats)
                lock.unlock()
                group.leave()
            }
        }

        group.notify(queue: .main) { [weak self] in
            guard let self else { return }
            self.webrtcStatsRequestInFlight = false
            self.lastWebRtcStatsPollAtMs = self.clock.nowMs()
            self.onStatsUpdated(Self.mergeRealtimeStats(stats))
        }
    }

    static func mergeRealtimeStats(_ stats: [RealtimeCallStats]) -> RealtimeCallStats {
        guard !stats.isEmpty else { return .empty }

        func sumNonNil(_ values: [Double]) -> Double? {
            guard !values.isEmpty else { return nil }
            return values.reduce(0, +)
        }

        func sumNonNilInt64(_ values: [Int64]) -> Int64? {
            guard !values.isEmpty else { return nil }
            return values.reduce(0, +)
        }

        var merged = RealtimeCallStats.empty
        merged.transportPath = Array(Set(stats.compactMap(\.transportPath))).sorted().joined(separator: " | ")
        if merged.transportPath?.isEmpty == true {
            merged.transportPath = nil
        }
        merged.rttMs = stats.compactMap(\.rttMs).max()
        merged.availableOutgoingKbps = stats.compactMap(\.availableOutgoingKbps).min()
        merged.audioRxCodec = joinCodecMimeTypes(stats.compactMap(\.audioRxCodec))
        merged.audioTxCodec = joinCodecMimeTypes(stats.compactMap(\.audioTxCodec))
        merged.audioRxPacketLossPct = stats.compactMap(\.audioRxPacketLossPct).max()
        merged.audioTxPacketLossPct = stats.compactMap(\.audioTxPacketLossPct).max()
        merged.audioJitterMs = stats.compactMap(\.audioJitterMs).max()
        merged.audioPlayoutDelayMs = stats.compactMap(\.audioPlayoutDelayMs).max()
        merged.audioConcealedPct = stats.compactMap(\.audioConcealedPct).max()
        merged.audioRxKbps = sumNonNil(stats.compactMap(\.audioRxKbps))
        merged.audioTxKbps = sumNonNil(stats.compactMap(\.audioTxKbps))
        merged.videoRxCodec = joinCodecMimeTypes(stats.compactMap(\.videoRxCodec))
        merged.videoTxCodec = joinCodecMimeTypes(stats.compactMap(\.videoTxCodec))
        merged.videoRxPacketLossPct = stats.compactMap(\.videoRxPacketLossPct).max()
        merged.videoTxPacketLossPct = stats.compactMap(\.videoTxPacketLossPct).max()
        merged.videoRxKbps = sumNonNil(stats.compactMap(\.videoRxKbps))
        merged.videoTxKbps = sumNonNil(stats.compactMap(\.videoTxKbps))
        merged.videoFps = stats.compactMap(\.videoFps).min()
        let resolutions = Array(Set(stats.compactMap(\.videoResolution))).sorted()
        merged.videoResolution = resolutions.isEmpty ? nil : resolutions.joined(separator: " | ")
        merged.videoFreezeCount60s = stats.compactMap(\.videoFreezeCount60s).reduce(0, +)
        merged.videoFreezeDuration60s = sumNonNil(stats.compactMap(\.videoFreezeDuration60s))
        merged.videoRetransmitPct = stats.compactMap(\.videoRetransmitPct).max()
        merged.videoNackPerMin = sumNonNil(stats.compactMap(\.videoNackPerMin))
        merged.videoPliPerMin = sumNonNil(stats.compactMap(\.videoPliPerMin))
        merged.videoFirPerMin = sumNonNil(stats.compactMap(\.videoFirPerMin))
        // Sum cumulative counters across slots (nil when
        // no slot reported the kind). CallQualityTracker / the host segment
        // diff rebaseline when the *aggregate* delta goes negative (counter
        // reset / slot teardown). In the dominant 1:1 case there is a single
        // slot so this is exact; for mesh, a simultaneous reset of one slot
        // masked by another slot's increase is an accepted approximation.
        merged.videoFramesDecoded = sumNonNilInt64(stats.compactMap(\.videoFramesDecoded))
        merged.videoFramesDropped = sumNonNilInt64(stats.compactMap(\.videoFramesDropped))
        merged.audioPacketsLost = sumNonNilInt64(stats.compactMap(\.audioPacketsLost))
        merged.audioPacketsReceived = sumNonNilInt64(stats.compactMap(\.audioPacketsReceived))
        merged.updatedAtMs = stats.map(\.updatedAtMs).max() ?? 0
        return merged
    }
}
